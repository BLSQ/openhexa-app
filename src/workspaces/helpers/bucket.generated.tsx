import * as Types from '../../graphql-types';

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
const defaultOptions = {} as const;
export type GetFileDownloadUrlMutationVariables = Types.Exact<{
  input: Types.PrepareObjectDownloadInput;
}>;


export type GetFileDownloadUrlMutation = { __typename?: 'Mutation', prepareObjectDownload: { __typename?: 'PrepareObjectDownloadResult', success: boolean, downloadUrl?: any | null } };

export type DeleteBucketObjectMutationVariables = Types.Exact<{
  input: Types.DeleteBucketObjectInput;
}>;


export type DeleteBucketObjectMutation = { __typename?: 'Mutation', deleteBucketObject: { __typename?: 'DeleteBucketObjectResult', success: boolean, errors: Array<Types.DeleteBucketObjectError> } };

export type GetBucketUploadUrlMutationVariables = Types.Exact<{
  input: Types.PrepareObjectUploadInput;
}>;


export type GetBucketUploadUrlMutation = { __typename?: 'Mutation', prepareObjectUpload: { __typename?: 'PrepareObjectUploadResult', success: boolean, uploadUrl?: any | null } };

export type CreateBucketFolderMutationVariables = Types.Exact<{
  input: Types.CreateBucketFolderInput;
}>;


export type CreateBucketFolderMutation = { __typename?: 'Mutation', createBucketFolder: { __typename?: 'CreateBucketFolderResult', success: boolean, errors: Array<Types.CreateBucketFolderError>, folder?: { __typename?: 'BucketObject', key: string, name: string, type: Types.BucketObjectType } | null } };


export const GetFileDownloadUrlDocument = gql`
    mutation GetFileDownloadUrl($input: PrepareObjectDownloadInput!) {
  prepareObjectDownload(input: $input) {
    success
    downloadUrl
  }
}
    `;
export type GetFileDownloadUrlMutationFn = Apollo.MutationFunction<GetFileDownloadUrlMutation, GetFileDownloadUrlMutationVariables>;

/**
 * __useGetFileDownloadUrlMutation__
 *
 * To run a mutation, you first call `useGetFileDownloadUrlMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useGetFileDownloadUrlMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [getFileDownloadUrlMutation, { data, loading, error }] = useGetFileDownloadUrlMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useGetFileDownloadUrlMutation(baseOptions?: Apollo.MutationHookOptions<GetFileDownloadUrlMutation, GetFileDownloadUrlMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<GetFileDownloadUrlMutation, GetFileDownloadUrlMutationVariables>(GetFileDownloadUrlDocument, options);
      }
export type GetFileDownloadUrlMutationHookResult = ReturnType<typeof useGetFileDownloadUrlMutation>;
export type GetFileDownloadUrlMutationResult = Apollo.MutationResult<GetFileDownloadUrlMutation>;
export type GetFileDownloadUrlMutationOptions = Apollo.BaseMutationOptions<GetFileDownloadUrlMutation, GetFileDownloadUrlMutationVariables>;
export const DeleteBucketObjectDocument = gql`
    mutation deleteBucketObject($input: DeleteBucketObjectInput!) {
  deleteBucketObject(input: $input) {
    success
    errors
  }
}
    `;
export type DeleteBucketObjectMutationFn = Apollo.MutationFunction<DeleteBucketObjectMutation, DeleteBucketObjectMutationVariables>;

/**
 * __useDeleteBucketObjectMutation__
 *
 * To run a mutation, you first call `useDeleteBucketObjectMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useDeleteBucketObjectMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [deleteBucketObjectMutation, { data, loading, error }] = useDeleteBucketObjectMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useDeleteBucketObjectMutation(baseOptions?: Apollo.MutationHookOptions<DeleteBucketObjectMutation, DeleteBucketObjectMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<DeleteBucketObjectMutation, DeleteBucketObjectMutationVariables>(DeleteBucketObjectDocument, options);
      }
export type DeleteBucketObjectMutationHookResult = ReturnType<typeof useDeleteBucketObjectMutation>;
export type DeleteBucketObjectMutationResult = Apollo.MutationResult<DeleteBucketObjectMutation>;
export type DeleteBucketObjectMutationOptions = Apollo.BaseMutationOptions<DeleteBucketObjectMutation, DeleteBucketObjectMutationVariables>;
export const GetBucketUploadUrlDocument = gql`
    mutation GetBucketUploadUrl($input: PrepareObjectUploadInput!) {
  prepareObjectUpload(input: $input) {
    success
    uploadUrl
  }
}
    `;
export type GetBucketUploadUrlMutationFn = Apollo.MutationFunction<GetBucketUploadUrlMutation, GetBucketUploadUrlMutationVariables>;

/**
 * __useGetBucketUploadUrlMutation__
 *
 * To run a mutation, you first call `useGetBucketUploadUrlMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useGetBucketUploadUrlMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [getBucketUploadUrlMutation, { data, loading, error }] = useGetBucketUploadUrlMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useGetBucketUploadUrlMutation(baseOptions?: Apollo.MutationHookOptions<GetBucketUploadUrlMutation, GetBucketUploadUrlMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<GetBucketUploadUrlMutation, GetBucketUploadUrlMutationVariables>(GetBucketUploadUrlDocument, options);
      }
export type GetBucketUploadUrlMutationHookResult = ReturnType<typeof useGetBucketUploadUrlMutation>;
export type GetBucketUploadUrlMutationResult = Apollo.MutationResult<GetBucketUploadUrlMutation>;
export type GetBucketUploadUrlMutationOptions = Apollo.BaseMutationOptions<GetBucketUploadUrlMutation, GetBucketUploadUrlMutationVariables>;
export const CreateBucketFolderDocument = gql`
    mutation CreateBucketFolder($input: CreateBucketFolderInput!) {
  createBucketFolder(input: $input) {
    success
    errors
    folder {
      key
      name
      type
    }
  }
}
    `;
export type CreateBucketFolderMutationFn = Apollo.MutationFunction<CreateBucketFolderMutation, CreateBucketFolderMutationVariables>;

/**
 * __useCreateBucketFolderMutation__
 *
 * To run a mutation, you first call `useCreateBucketFolderMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useCreateBucketFolderMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [createBucketFolderMutation, { data, loading, error }] = useCreateBucketFolderMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useCreateBucketFolderMutation(baseOptions?: Apollo.MutationHookOptions<CreateBucketFolderMutation, CreateBucketFolderMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<CreateBucketFolderMutation, CreateBucketFolderMutationVariables>(CreateBucketFolderDocument, options);
      }
export type CreateBucketFolderMutationHookResult = ReturnType<typeof useCreateBucketFolderMutation>;
export type CreateBucketFolderMutationResult = Apollo.MutationResult<CreateBucketFolderMutation>;
export type CreateBucketFolderMutationOptions = Apollo.BaseMutationOptions<CreateBucketFolderMutation, CreateBucketFolderMutationVariables>;