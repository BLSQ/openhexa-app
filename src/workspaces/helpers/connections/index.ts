import { ConnectionType } from "graphql-types";
export { convertFieldsToInput } from "./utils";
export type { ConnectionForm, FieldForm } from "./utils";

import postgresql from "./postgresql";
import { ConnectionDefinition } from "./utils";
import custom from "./custom";
import s3 from "./s3";
import gcs from "./gcs";
import iaso from "./iaso";
import dhis2 from "./dhis2";

const Connections: { [key in ConnectionType]: ConnectionDefinition } = {
  [ConnectionType.Postgresql]: postgresql,
  [ConnectionType.Custom]: custom,
  [ConnectionType.S3]: s3,
  [ConnectionType.Gcs]: gcs,
  [ConnectionType.Iaso]: iaso,
  [ConnectionType.Dhis2]: dhis2,
};

export default Connections;
