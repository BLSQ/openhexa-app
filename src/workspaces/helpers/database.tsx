type lang = "R" | "PYTHON";

export const getReadTableSnippet = (tableName: string) => {
  return `import os
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(os.environ["WORKSPACE_DATABASE_URL"])

pd.read_sql("SELECT * FROM ${tableName}", con=engine)`;
};

export const getUsageSnippet = (tableName: string, lang: lang) => {
  let text = "";
  switch (lang) {
    case "PYTHON":
      text = `import os
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(os.environ["WORKSPACE_DATABASE_URL"])

# Create sample dataframe
df = pd.DataFrame({"name": ["Jane", "John", "Tyler"], "age": [19, 17, 22]})

# Write data
df.to_sql("${tableName}", con=engine, if_exists="replace")`;
      break;
    case "R":
      text = `library(DBI)
  
con <- dbConnect(
    RPostgres::Postgres(),
    dbname = Sys.getenv("WORKSPACE_DATABASE_DB_NAME"),
    host = Sys.getenv("WORKSPACE_DATABASE_HOST"),
    port = Sys.getenv("WORKSPACE_DATABASE_PORT"),
    user = Sys.getenv("WORKSPACE_DATABASE_USERNAME"),
    password = Sys.getenv("WORKSPACE_DATABASE_PASSWORD")
)

dbWriteTable(con, "${tableName}", Data_fin, overwrite=TRUE)`;
      break;
  }
  return text;
};
