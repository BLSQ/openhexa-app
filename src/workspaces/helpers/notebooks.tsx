import { ApolloClient, gql } from "@apollo/client";
import {
  LaunchNotebookServerMutation,
  LaunchNotebookServerMutationVariables,
} from "./notebooks.generated";

export async function launchNotebookServer(
  client: ApolloClient<unknown>,
  workspaceSlug: string
) {
  const { data } = await client.mutate<
    LaunchNotebookServerMutation,
    LaunchNotebookServerMutationVariables
  >({
    mutation: gql`
      mutation launchNotebookServer($input: LaunchNotebookServerInput!) {
        launchNotebookServer(input: $input) {
          success
          server {
            name
            ready
            url
          }
        }
      }
    `,
    variables: {
      input: {
        workspaceSlug,
      },
    },
  });

  if (!data?.launchNotebookServer.success) {
    throw new Error(`Unable to launch notebook server for ${workspaceSlug}`);
  }

  return data.launchNotebookServer.server!;
}
