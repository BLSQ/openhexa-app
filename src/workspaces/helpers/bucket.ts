import { gql } from "@apollo/client";
import { getApolloClient } from "core/helpers/apollo";
import {
  GetFileDownloadUrlMutation,
  GetFileDownloadUrlMutationVariables,
  DeleteBucketObjectMutation,
  DeleteBucketObjectMutationVariables,
  GetBucketUploadUrlMutation,
  GetBucketUploadUrlMutationVariables,
} from "./bucket.generated";

export async function getBucketObjectDownloadUrl(
  workspaceSlug: string,
  objectKey: string,
): Promise<string> {
  const client = getApolloClient();
  const { data } = await client.mutate<
    GetFileDownloadUrlMutation,
    GetFileDownloadUrlMutationVariables
  >({
    mutation: gql`
      mutation GetFileDownloadUrl($input: PrepareObjectDownloadInput!) {
        prepareObjectDownload(input: $input) {
          success
          downloadUrl
        }
      }
    `,
    variables: {
      input: {
        workspaceSlug,
        objectKey,
      },
    },
  });

  if (data?.prepareObjectDownload?.success) {
    return data.prepareObjectDownload.downloadUrl as string;
  } else {
    throw new Error("Object cannot be downloaded");
  }
}

export async function downloadURL(url: string, target: string = "") {
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.target = target;
  document.body.appendChild(anchor);
  anchor.click();
  document.body.removeChild(anchor);
}

export async function deleteBucketObject(
  workspaceSlug: string,
  objectKey: string,
) {
  const client = getApolloClient();
  const { data } = await client.mutate<
    DeleteBucketObjectMutation,
    DeleteBucketObjectMutationVariables
  >({
    mutation: gql`
      mutation deleteBucketObject($input: DeleteBucketObjectInput!) {
        deleteBucketObject(input: $input) {
          success
          errors
        }
      }
    `,
    variables: { input: { workspaceSlug, objectKey } },
    update(cache) {
      const normalizedId = cache.identify({
        key: objectKey,
        __typename: "BucketObject",
      });

      cache.evict({ id: normalizedId });
      cache.gc();
    },
  });

  if (!data?.deleteBucketObject.success) {
    throw new Error("Unexpected error.");
  }
}

export async function getBucketObjectUploadUrl(
  workspaceSlug: string,
  key: string,
  contentType: string,
) {
  const client = getApolloClient();
  const { data } = await client.mutate<
    GetBucketUploadUrlMutation,
    GetBucketUploadUrlMutationVariables
  >({
    mutation: gql`
      mutation GetBucketUploadUrl($input: PrepareObjectUploadInput!) {
        prepareObjectUpload(input: $input) {
          success
          uploadUrl
        }
      }
    `,
    variables: {
      input: {
        workspaceSlug,
        objectKey: key,
        contentType,
      },
    },
  });

  if (data?.prepareObjectUpload?.success) {
    return data.prepareObjectUpload.uploadUrl as string;
  } else {
    throw new Error("Object cannot be uploaded");
  }
}

export async function createBucketFolder(
  workspaceSlug: string,
  folderKey: string,
) {
  const client = getApolloClient();
  const { data } = await client.mutate({
    mutation: gql`
      mutation CreateBucketFolder($input: CreateBucketFolderInput!) {
        createBucketFolder(input: $input) {
          success
          errors
          folder {
            key
            name
            type
          }
        }
      }
    `,
    variables: {
      input: {
        workspaceSlug,
        folderKey,
      },
    },
  });

  if (!data?.createBucketFolder.success) {
    throw new Error("Unexpected error.");
  }

  return data.createBucketFolder.folder;
}
