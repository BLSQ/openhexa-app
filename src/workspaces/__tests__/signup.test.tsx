import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {
  JoinWorkspaceDocument,
  useJoinWorkspaceMutation,
} from "workspaces/graphql/mutations.generated";
import { TestApp } from "core/helpers/testutils";
import mockRouter from "next-router-mock";
import router from "next/router";
import SignUpPage from "pages/workspaces/[workspaceSlug]/signup";
import { faker } from "@faker-js/faker";

const USER = {
  email: faker.internet.email(),
  firstName: faker.name.firstName(),
  lastName: faker.name.lastName(),
  token: faker.datatype.string(20),
};

jest.mock("workspaces/graphql/mutations.generated", () => {
  const actualModule = jest.requireActual(
    "workspaces/graphql/mutations.generated"
  );
  return {
    ...actualModule,
    __esModule: true,
    useJoinWorkspaceMutation: jest.fn(actualModule.useJoinWorkspaceMutation),
  };
});

//

const useJoinWorkspaceMutationMock = useJoinWorkspaceMutation as jest.Mock;
describe("WorkspaceSignUpPage", () => {
  beforeEach(() => {
    mockRouter.setCurrentUrl("/");
    useJoinWorkspaceMutationMock.mockClear();
  });

  it("renders the sign up page", async () => {
    const { container } = render(
      <TestApp>
        <SignUpPage />
      </TestApp>
    );

    expect(container).toMatchSnapshot();
  });

  it("calls the join workspace action when form is correctly filled", async () => {
    const joinWorkspace = jest.fn();
    useJoinWorkspaceMutationMock.mockReturnValue([joinWorkspace]);
    const user = userEvent.setup();

    render(
      <TestApp>
        <SignUpPage email={USER.email} token={USER.token} />
      </TestApp>
    );

    const submitBtn = screen.getByRole("button", { name: "Submit" });

    await user.click(submitBtn);
    expect(joinWorkspace).not.toHaveBeenCalled();

    const firstName = screen.getByTestId("firstName");
    await user.type(firstName, USER.firstName);
    await user.click(submitBtn);
    expect(joinWorkspace).not.toHaveBeenCalled();

    const lastName = screen.getByTestId("lastName");
    await user.type(lastName, USER.lastName);
    await user.click(submitBtn);
    expect(joinWorkspace).not.toHaveBeenCalled();

    const password = screen.getByTestId("password");
    await user.type(password, "pA$$W0rd");
    await user.click(submitBtn);
    expect(joinWorkspace).not.toHaveBeenCalled();

    const confirmPassword = screen.getByTestId("confirmPassword");
    await user.type(confirmPassword, "foo@bar");
    await user.click(submitBtn);

    expect(joinWorkspace).not.toHaveBeenCalled();
    const errorMessage = await screen.findByText(
      "The password confirmation does not match the given password"
    );
    expect(errorMessage).toBeInTheDocument();

    await user.clear(confirmPassword);
    await user.type(confirmPassword, "pA$$W0rd");
    await user.click(submitBtn);

    expect(joinWorkspace).toHaveBeenCalled();
    expect(joinWorkspace).toHaveBeenCalledWith({
      variables: {
        input: {
          firstName: USER.firstName,
          lastName: USER.lastName,
          token: USER.token,
          password: "pA$$W0rd",
          confirmPassword: "pA$$W0rd",
        },
      },
    });
  });

  it("redirects the user on success", async () => {
    const pushSpy = jest.spyOn(router, "push");
    const { useJoinWorkspaceMutation } = jest.requireActual(
      "workspaces/graphql/mutations.generated"
    );
    useJoinWorkspaceMutationMock.mockImplementation(useJoinWorkspaceMutation);
    const user = userEvent.setup();
    const mocks = [
      {
        request: {
          query: JoinWorkspaceDocument,
          variables: {
            input: {
              firstName: USER.firstName,
              lastName: USER.lastName,
              token: USER.token,
              password: "pA$$W0rd",
              confirmPassword: "pA$$W0rd",
            },
          },
        },
        result: {
          data: {
            joinWorkspace: {
              success: true,
              errors: [],
              workspace: {
                slug: "slug",
              },
            },
          },
        },
      },
    ];

    render(
      <TestApp mocks={mocks}>
        <SignUpPage email={USER.email} token={USER.token} />
      </TestApp>
    );

    const submitBtn = screen.getByRole("button", { name: "Submit" });

    await user.click(submitBtn);
    expect(pushSpy).not.toHaveBeenCalled();

    const firstName = screen.getByTestId("firstName");
    const lastName = screen.getByTestId("lastName");
    const password = screen.getByTestId("password");
    const confirmPassword = screen.getByTestId("confirmPassword");

    await user.type(firstName, USER.firstName);
    await user.type(lastName, USER.lastName);
    await user.type(password, "pA$$W0rd");
    await user.type(confirmPassword, "pA$$W0rd");

    await user.click(submitBtn);

    waitFor(() => {
      expect(pushSpy).toHaveBeenCalled();
    });
  });

  it("displays invalid password formet error message when the password is similar to user info", async () => {
    const pushSpy = jest.spyOn(router, "push");
    const { useJoinWorkspaceMutation } = jest.requireActual(
      "workspaces/graphql/mutations.generated"
    );
    useJoinWorkspaceMutationMock.mockImplementation(useJoinWorkspaceMutation);
    const user = userEvent.setup();
    const mocks = [
      {
        request: {
          query: JoinWorkspaceDocument,
          variables: {
            input: {
              firstName: USER.firstName,
              lastName: USER.lastName,
              token: USER.token,
              password: USER.firstName,
              confirmPassword: USER.firstName,
            },
          },
        },
        result: {
          data: {
            joinWorkspace: {
              success: false,
              errors: ["INVALID_CREDENTIALS"],
              workspace: null,
            },
          },
        },
      },
    ];

    render(
      <TestApp mocks={mocks}>
        <SignUpPage email={USER.email} token={USER.token} />
      </TestApp>
    );

    const submitBtn = screen.getByRole("button", { name: "Submit" });

    await user.click(submitBtn);
    expect(pushSpy).not.toHaveBeenCalled();

    const firstName = screen.getByTestId("firstName");
    const lastName = screen.getByTestId("lastName");
    const password = screen.getByTestId("password");
    const confirmPassword = screen.getByTestId("confirmPassword");

    await user.type(firstName, USER.firstName);
    await user.type(lastName, USER.lastName);
    await user.type(password, USER.firstName);
    await user.type(confirmPassword, USER.firstName);
    await user.click(submitBtn);

    const errorMessage = await screen.findByText("Invalid password format");
    expect(errorMessage).toBeInTheDocument();
  });

  it("displays an invalid token error message when the token format is not correct or not linked to an invitation", async () => {
    const pushSpy = jest.spyOn(router, "push");
    const { useJoinWorkspaceMutation } = jest.requireActual(
      "workspaces/graphql/mutations.generated"
    );
    useJoinWorkspaceMutationMock.mockImplementation(useJoinWorkspaceMutation);
    const user = userEvent.setup();
    const mocks = [
      {
        request: {
          query: JoinWorkspaceDocument,
          variables: {
            input: {
              firstName: USER.firstName,
              lastName: USER.lastName,
              token: "123",
              password: "pA$$W0rd",
              confirmPassword: "pA$$W0rd",
            },
          },
        },
        result: {
          data: {
            joinWorkspace: {
              success: false,
              errors: ["INVALID_TOKEN"],
              workspace: null,
            },
          },
        },
      },
    ];

    render(
      <TestApp mocks={mocks}>
        <SignUpPage email={USER.email} token={"123"} />
      </TestApp>
    );

    const submitBtn = screen.getByRole("button", { name: "Submit" });

    await user.click(submitBtn);
    expect(pushSpy).not.toHaveBeenCalled();

    const firstName = screen.getByTestId("firstName");
    const lastName = screen.getByTestId("lastName");
    const password = screen.getByTestId("password");
    const confirmPassword = screen.getByTestId("confirmPassword");

    await user.type(firstName, USER.firstName);
    await user.type(lastName, USER.lastName);
    await user.type(password, "pA$$W0rd");
    await user.type(confirmPassword, "pA$$W0rd");
    await user.click(submitBtn);

    const errorMessage = await screen.findByText("The invite link is invalid.");
    expect(errorMessage).toBeInTheDocument();
  });

  it("displays expired invitation error message when the token is expired", async () => {
    const pushSpy = jest.spyOn(router, "push");
    const { useJoinWorkspaceMutation } = jest.requireActual(
      "workspaces/graphql/mutations.generated"
    );
    useJoinWorkspaceMutationMock.mockImplementation(useJoinWorkspaceMutation);
    const user = userEvent.setup();
    const mocks = [
      {
        request: {
          query: JoinWorkspaceDocument,
          variables: {
            input: {
              firstName: USER.firstName,
              lastName: USER.lastName,
              token: USER.token,
              password: "pA$$W0rd",
              confirmPassword: "pA$$W0rd",
            },
          },
        },
        result: {
          data: {
            joinWorkspace: {
              success: false,
              errors: ["EXPIRED_TOKEN"],
              workspace: null,
            },
          },
        },
      },
    ];

    render(
      <TestApp mocks={mocks}>
        <SignUpPage email={USER.email} token={USER.token} />
      </TestApp>
    );

    const submitBtn = screen.getByRole("button", { name: "Submit" });

    await user.click(submitBtn);
    expect(pushSpy).not.toHaveBeenCalled();

    const firstName = screen.getByTestId("firstName");
    const lastName = screen.getByTestId("lastName");
    const password = screen.getByTestId("password");
    const confirmPassword = screen.getByTestId("confirmPassword");

    await user.type(firstName, USER.firstName);
    await user.type(lastName, USER.lastName);
    await user.type(password, "pA$$W0rd");
    await user.type(confirmPassword, "pA$$W0rd");
    await user.click(submitBtn);

    const errorMessage = await screen.findByText("The invitation has expired.");
    expect(errorMessage).toBeInTheDocument();
  });

  it("displays account already exists error message when the user has already an account", async () => {
    const pushSpy = jest.spyOn(router, "push");
    const { useJoinWorkspaceMutation } = jest.requireActual(
      "workspaces/graphql/mutations.generated"
    );
    useJoinWorkspaceMutationMock.mockImplementation(useJoinWorkspaceMutation);
    const user = userEvent.setup();
    const mocks = [
      {
        request: {
          query: JoinWorkspaceDocument,
          variables: {
            input: {
              firstName: USER.firstName,
              lastName: USER.lastName,
              token: USER.token,
              password: "pA$$W0rd",
              confirmPassword: "pA$$W0rd",
            },
          },
        },
        result: {
          data: {
            joinWorkspace: {
              success: false,
              errors: ["ALREADY_EXISTS"],
              workspace: null,
            },
          },
        },
      },
    ];

    render(
      <TestApp mocks={mocks}>
        <SignUpPage email={USER.email} token={USER.token} />
      </TestApp>
    );

    const submitBtn = screen.getByRole("button", { name: "Submit" });

    await user.click(submitBtn);
    expect(pushSpy).not.toHaveBeenCalled();

    const firstName = screen.getByTestId("firstName");
    const lastName = screen.getByTestId("lastName");
    const password = screen.getByTestId("password");
    const confirmPassword = screen.getByTestId("confirmPassword");

    await user.type(firstName, USER.firstName);
    await user.type(lastName, USER.lastName);
    await user.type(password, "pA$$W0rd");
    await user.type(confirmPassword, "pA$$W0rd");
    await user.click(submitBtn);

    const errorMessage = await screen.findByText(
      "An account already exists with this email address. Please go to the login page."
    );
    expect(errorMessage).toBeInTheDocument();
  });
});
