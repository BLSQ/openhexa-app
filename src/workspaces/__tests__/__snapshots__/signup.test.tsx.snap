// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkspaceSignUpPage renders the sign up page 1`] = `
<div>
  <form
    class="max-w-md flex-1 space-y-6"
  >
    <div>
      <div
        class="relative h-16 w-auto"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="OpenHEXA logo"
            class="mx-auto block h-16 w-auto"
            data-nimg="fill"
            decoding="async"
            sizes="100vw"
            src="/images/logo.svg"
            srcset="/images/logo.svg 640w, /images/logo.svg 750w, /images/logo.svg 828w, /images/logo.svg 1080w, /images/logo.svg 1200w, /images/logo.svg 1920w, /images/logo.svg 2048w, /images/logo.svg 3840w"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </span>
      </div>
      <h2
        class="mt-6 text-center text-3xl font-extrabold text-gray-900"
      >
        Sign up
      </h2>
    </div>
    <div
      class="space-y-4 pt-2"
    >
      <div>
        <div
          class="mb-1 flex justify-between"
        >
          <label
            class="block text-sm font-medium text-gray-600 flex"
            for="email"
          >
            Email
          </label>
        </div>
        <div
          class="group relative"
        >
          <input
            aria-describedby="email-description"
            aria-invalid="false"
            class="w-full form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500"
            disabled=""
            id="email"
            name="email"
            required=""
            type="email"
            value=""
          />
        </div>
      </div>
      <div>
        <div
          class="mb-1 flex justify-between"
        >
          <label
            class="block text-sm font-medium text-gray-600 flex"
            for="firstName"
          >
            First name
          </label>
        </div>
        <div
          class="group relative"
        >
          <input
            aria-describedby="firstName-description"
            aria-invalid="false"
            class="w-full form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500"
            data-testid="firstName"
            id="firstName"
            name="firstName"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
      <div>
        <div
          class="mb-1 flex justify-between"
        >
          <label
            class="block text-sm font-medium text-gray-600 flex"
            for="lastName"
          >
            Last name
          </label>
        </div>
        <div
          class="group relative"
        >
          <input
            aria-describedby="lastName-description"
            aria-invalid="false"
            class="w-full form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500"
            data-testid="lastName"
            id="lastName"
            name="lastName"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
      <div>
        <div>
          <div
            class="mb-1 flex justify-between"
          >
            <label
              class="block text-sm font-medium text-gray-600 flex"
              for="password"
            >
              Password
            </label>
          </div>
          <div
            class="group relative"
          >
            <input
              aria-describedby="password-description"
              aria-invalid="false"
              class="w-full form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500"
              data-testid="password"
              id="password"
              name="password"
              required=""
              type="password"
              value=""
            />
          </div>
        </div>
      </div>
      <div>
        <div
          class="mb-1 flex justify-between"
        >
          <label
            class="block text-sm font-medium text-gray-600 flex"
            for="confirmPassword"
          >
            Confirm password
          </label>
        </div>
        <div
          class="group relative"
        >
          <input
            aria-describedby="confirmPassword-description"
            aria-invalid="false"
            class="w-full form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500"
            data-testid="confirmPassword"
            id="confirmPassword"
            name="confirmPassword"
            required=""
            type="password"
            value=""
          />
        </div>
      </div>
      <p
        class="mt-2 text-xs"
      >
        <span>
          The password must respect the following criterias:
        </span>
      </p>
      <ul
        class="list my-2 list-inside list-disc text-xs"
      >
        <li>
          At least 8 characters.
        </li>
        <li>
          Alphanumeric (can't be entirely numeric)
          .
        </li>
        <li>
          Can't be a commonly used password or similar to your personal information
        </li>
      </ul>
    </div>
    <div
      class="space-y-2"
    >
      <button
        class="relative rounded h-fit
    focus:outline-none focus:ring-2 focus:ring-offset-2
    inline-flex items-center justify-center transition-all
    border font-medium
    disabled:opacity-50 disabled:cursor-not-allowed
   text-sm px-3 py-2.5 leading-4 shadow-sm border-transparent text-white bg-blue-600 hover:bg-blue-700 focus:ring-blue-500 w-full"
        data-testid="submit"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
